cmakelib/lexer/filter.go

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright 2019 The Kythe Authors. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package lexer
+
+import (
+	"github.com/alecthomas/participle/lexer"
+)
+
+// filterLexer processes the raw stream of tokens coming from the underlying scanner-based lexer
+// to make the resulting AST simpler and more readable.
+type filterLexer struct {
+	l    lexer.Lexer
+	buf  []lexer.Token
+	prev lexer.Token
+}
+
+// Next implements the lexer.Lexer interface for filterLexer.
+func (l *filterLexer) Next() (lexer.Token, error) {
+	var err error
+	if len(l.buf) > 0 {
+		tok := l.buf[0]
+		l.buf = l.buf[1:]
+		return tok, nil
+	}
+	switch l.prev.Type {
+	case BracketOpen:
+		return l.bufferTokens(combineBracketContent(l.l, len(l.prev.Value)))
+	case Quote:
+		return l.bufferTokens(combineQuotedContent(l.l))
+	}
+scan:
+	for {
+		l.prev, err = l.l.Next()
+		if err != nil {
+			break
+		}
+		switch l.prev.Type {
+		case Comment:
+			l.prev, err = l.bufferTokens(consumeComment(l.l))
+			if err != nil || l.prev.Type == lexer.EOF {
+				break scan
+			}
+		default:
+			break scan
+		}
+	}
+	if err != nil {
+		l.prev = lexer.Token{}
+	}
+	return l.prev, err
+}
+
+// bufferTokens updates the buffer, clears tok and returns the first buffered token or error.
+func (l *filterLexer) bufferTokens(toks []lexer.Token, done bool, err error) (lexer.Token, error) {
+	if done || err != nil {
+		l.prev = lexer.Token{}
+	}
+	if len(toks) > 0 {
+		l.buf = toks[1:]
+		return toks[0], err
+	}
+	l.buf = nil
+	return lexer.Token{}, err
+}
+
+// combineBracketContent merges tokens from the lexer until it encounters
+// a BracketClose token with a value equal to hdrlen, EOF or an error.
+// The terminating bracket is included in the returned value.
+func combineBracketContent(l lexer.Lexer, hdrlen int) ([]lexer.Token, bool, error) {
```

Comment:
Can you give a blurb about the returned values? Specifically, what the second value (the bool) is? It looks like it is always true.

Similar comment for the following 2 functions.

Reply:
Done. Specifically, these are all fed directly into `bufferTokens` and the bool signals whether or not the combining sequence is "complete".  This enables the optimization creachadair commented on earlier where we don't have to consume all the way to the end of the string, just to the next uncombined token.